### PR TITLE
[PluggableDevice] Add pluggable profiler C Interface support

### PR DESCRIPTION
**Pull request by jzhoulon**
Add TensorFlow profiler support for PluggableDevice.

### CODE REVIEW

**Comment by jzhoulon** [comment]
@penpornk @yisitu @jbaiocchi This PR is the implementation of the RFC: [Add pluggable device profiler support ](https://github.com/tensorflow/community/pull/389), please help to review. Thanks very much

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/profiler.cc:117]: Add 'override' to functions that implement ProfilerInterface.
https://google.github.io/styleguide/cppguide.html#Inheritance

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/profiler.cc:152]: This function can be removed, RunMetadata overload of CollectData has been removed from ProfilerInterface.

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/profiler.cc:131]: into it ?

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/profiler.cc:220]: std::move(create_func)

File: tensorflow/c/experimental/profiler/BUILD
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/BUILD:22]: Can you rename this directory and all the files to be "pluggable_profiler" instead of just "profiler"?

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/profiler.cc:31]: Is this used? If not remove.

File: tensorflow/c/experimental/profiler/profiler_internal.h
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/profiler_internal.h:25]: This can be removed.

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/profiler.cc:118]: The pattern I see elsewhere in the code base is:

std::unique_ptr<TF_Status, decltype(&TF_DeleteStatus)> status(
      TF_NewStatus(), TF_DeleteStatus);

I would also name this variable status instead of c_status, the prefix does not add much value in this case.

File: tensorflow/c/experimental/profiler/profiler_internal.h
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/profiler_internal.h:29]: Rewrite as "using TFInitProfilerFn = ".
https://google.github.io/styleguide/cppguide.html#Aliases

File: tensorflow/c/experimental/profiler/profiler_internal.h
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/profiler_internal.h:32]: Add a black-box documentation comment to explain what this function does.

File: tensorflow/c/experimental/profiler/profiler.h
Comment (jbaiocchi) [inline tensorflow/c/experimental/profiler/profiler.h:31]: Add documentation comments for each of the structs.

File: tensorflow/core/common_runtime/pluggable_device/pluggable_device_plugin_init.cc
Comment (jbaiocchi) [inline tensorflow/core/common_runtime/pluggable_device/pluggable_device_plugin_init.cc:104]: A returned Status must not be ignored. If not using TF_RETURN_IF_ERROR, explicity call IgnoreError() on the result of the call and explain in a comment why it is OK to ignore. Even better, log the resulting status so for ease of debugging.

**Review by jbaiocchi — CHANGES_REQUESTED** [review]

File: tensorflow/c/experimental/profiler/BUILD
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/BUILD]: done

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/profiler.cc]: done

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/profiler.cc]: done

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/profiler.cc]: done

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/profiler.cc]: done

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/profiler.cc]: Done

File: tensorflow/c/experimental/profiler/profiler.cc
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/profiler.cc]: done

File: tensorflow/c/experimental/profiler/profiler.h
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/profiler.h]: done

File: tensorflow/c/experimental/profiler/profiler_internal.h
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/profiler_internal.h]: done

File: tensorflow/c/experimental/profiler/profiler_internal.h
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/profiler_internal.h]: done

File: tensorflow/c/experimental/profiler/profiler_internal.h
Comment (jzhoulon) [reply tensorflow/c/experimental/profiler/profiler_internal.h]: - done

File: tensorflow/core/common_runtime/pluggable_device/pluggable_device_plugin_init.cc
Comment (jzhoulon) [reply tensorflow/core/common_runtime/pluggable_device/pluggable_device_plugin_init.cc]: Done

**Comment by jzhoulon** [comment]
@jbaiocchi Thanks for the comments, I have addressed all the comments, please help review it. thanks very much

File: tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.h
Comment (jbaiocchi) [inline tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.h:50]: nit: add // even to these blank lines as for a single comment block

File: tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.cc
Comment (jbaiocchi) [inline tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.cc:72]: This is duplicated code from tensorflow/c/experimental/stream_executor/stream_executor.cc
Can we prevent duplication by first extracting a common library, e.g., pluggable_device_utils ?

File: tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.cc
Comment (jbaiocchi) [inline tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.cc:76]: This struct and the OwnedTFStatus alias are also used in third_party/tensorflow/c/experimental/stream_executor/stream_executor.cc

Move them to common utilities as well.

**Review by jbaiocchi — CHANGES_REQUESTED** [review]
Please address the code duplication issue.

**Review by jbaiocchi — APPROVED** [review]

File: tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.h
Comment (penpornk) [inline tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.h:1]: Let's wrap functions here in `namespace pluggable_device` to avoid symbol clashes.

File: tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.h
Comment (penpornk) [inline tensorflow/c/experimental/pluggable_profiler/pluggable_profiler.h:73]: Minor formatting nit:
```suggestion
//  void profiler_collect_data_xspace(const TP_Profiler* profiler, uint8_t*
//  buffer, size_t* size_in_bytes, TF_Status* status) {
//    /* Plugin generates Xspace based on collected profiler data. */
//    Xspace xspace = get_my_xspace();
//    size_t buffer_size_in_bytes = *size_in_bytes;
//    *size_in_bytes = xspace.ByteSizeLong(); /* get the size of Xspace */
//    if (buffer == nullptr) {
//      /* TensorFlow will first get the size of Xspace, then allocate the big
//         enough buffer and pass it to the plugin for retrieving Xspace. */
//      return;
//    }
//    bool success = xspace.SerializeToArray(buffer, buffer_size_in_bytes);
//  }
```

File: tensorflow/c/experimental/pluggable_profiler/pluggable_profiler_internal.h
Comment (penpornk) [inline tensorflow/c/experimental/pluggable_profiler/pluggable_profiler_internal.h:30]: Minor formatting nit:
Please add a newline before this line.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much for the PR! 

Would you mind adding simple mock tests in a follow-up PR? You can follow examples in [stream_executor_test.cc](
https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/stream_executor/stream_executor_test.cc). In our case we'll have mock start, stop, collect_data_xspace, etc. 

For this PR, I have a few minor formatting nits.

File: tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.cc
Comment (jzhoulon) [reply tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.cc:23]: done

File: tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.h
Comment (jzhoulon) [reply tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.h:1]: done

File: tensorflow/c/experimental/pluggable_profiler/pluggable_profiler_internal.h
Comment (jzhoulon) [reply tensorflow/c/experimental/pluggable_profiler/pluggable_profiler_internal.h:32]: done

**Comment by jzhoulon** [comment]
@penpornk Thanks for the comments, I have addressed all the comments, please help to review. And I will submit an follow-up PR to add mock test for pluggable profiler. Thanks.

**Review by penpornk — APPROVED** [review]
Thank you for changes! :)

**Comment by jzhoulon** [comment]
@penpornk I saw some internal checks failed, if there is some thing need to fix, please let me know, thanks very much.

**Comment by penpornk** [comment]
@jzhoulon Thanks for checking! I've taken care of it. Right now it's waiting for reviews from API owners. :)

File: tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.h
Comment (jzhoulon) [inline tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.h:24]: @penpornk The VALIDATA_STRUCT_SIZE seems be moved to [c_api_macros.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api_macros.h#L19-L21) after this PR merged, this will cause build failure in plugin. VALIDATE_STRUCT_SIZE is not a public C API(only used by stream_executor.cc and pluggable_profiler.cc). it contains some C++ struct(tensorflow::Status), so it will cause some issue when c_api_macros.h is included by plugin, I think this part of code needs to be moved to an TensorFlow internal files. from c_api_macros.h, Thanks very much.
build errror:
```
bazel-out/k8-opt/bin/external/local_config_tf/include/tensorflow/core/platform/status.h:287:65: note: passing argument to parameter 'v' here
inline tensorflow::string* TfCheckOpHelper(::tensorflow::Status v,
^
itex/core/graph/onednn_layout/onednn_layout.cc:263:15: error: no viable conversion from 'itex::Status' to '::tensorflow::Status'
TF_CHECK_OK(GetNodeAttr(node_def, "transpose_b", &trans_b));
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment (jbaiocchi) [reply tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.h:24]: It doesn't look related. Looks like you are calling TF_CHECK_OK, which is a TF macro that expands into a function (TfCheckOpHelper) that expects ::tensorflow::Status. GetNodeAttr should return ::tensorflow::Status but somehow the compiler thinks it is itex::Status. Note that VALIDATE_STRUCT_SIZE has always depended on ::tensorflow::Status. We added ifdef __cplusplus to ensure inclusion of c_api_macros into C code is OK.

Comment (jzhoulon) [reply tensorflow/c/experimental/pluggable_device_utils/pluggable_device_utils.h:24]: Thanks for the help! we find that we have duplicate name of macros in c_api_macros.h, we renamed them and it works successfully.

**Pull request merged**
